interpreter/cling/lib/Interpreter/DefinitionShadowing.cpp

```diff
@@ -0,0 +1,145 @@
+    LookupResult Previous(*m_Sema, D->getDeclName(), D->getLocation(),
+                   Sema::LookupOrdinaryName, Sema::ForRedeclaration);
+    m_Sema->LookupQualifiedName(Previous, m_TU);
+
+    for (auto I : Previous) {
+      if (I == D) continue;
+      auto NS = dyn_cast<NamespaceDecl>(I->getLexicalDeclContext());
+      if (isDefinition(I)
+          && (!isDefinition(D) || !(NS && NS->getName().startswith("__cling_N5")))) continue;
+
+      hideDecl(I);
+
+      if (TagDecl *TD = dyn_cast<TagDecl>(I)) {
+        if (!TD->isTransparentContext()) continue;
+        for (auto J : TD->decls()) hideDecl(cast<NamedDecl>(J));
+      }
+    }
+    return !Previous.empty();
+  }
+
+  /// \brief Invalidate previous function definition.  Local declararations might be
+  /// moved by DeclExtractor; in that case, invalidate all those before DeclExtractor runs.
+  void DefinitionShadowing::invalidatePreviousDefinitions(FunctionDecl *D) {
+    const CompilationOptions &CO = getTransaction()->getCompilationOpts();
+    if (CO.DeclarationExtraction) {
+      // XXX: this should make DeclExtractor behave as usual
+      auto CS = dyn_cast<CompoundStmt>(D->getBody());
+      for (auto I = CS->body_begin(), EI = CS->body_end(); I != EI; ++I) {
+        auto DS = dyn_cast<DeclStmt>(*I);
+        if (!DS) continue;
+
+        for (auto J = DS->decl_begin(); J != DS->decl_end(); ++J)
+          if (auto ND = dyn_cast<NamedDecl>(*J))
+            // Ignore any forward declaration issued after a definition. Fixes "error
+            // : reference to 'xxx' is ambiguous" in `class C {}; class C; C foo;`. 
+            if (invalidatePreviousDefinitions(ND) && !isDefinition(ND))
+              ND->setInvalidDecl();
+      }
+    } else
+      invalidatePreviousDefinitions(cast<NamedDecl>(D));
+  }
+
+  ASTTransformer::Result DefinitionShadowing::Transform(Decl* D) {
+    auto FD = dyn_cast<FunctionDecl>(D);
+    const CompilationOptions &CO = getTransaction()->getCompilationOpts();
+    // Declarations whose origin is the Cling prompt are subject to be nested in a
+    // `__cling_N5' namespace.
+    if (D->getLexicalDeclContext() != m_TU
+        // FIXME: NamespaceDecl/TemplateDecl require additional processing (TBD)
+        || isa<NamespaceDecl>(D) || D->isTemplateDecl() || (FD && FD->isFunctionTemplateSpecialization())
+        || isa<UsingDirectiveDecl>(D)
+        || !typedInClingPrompt(FullSourceLoc{D->getLocation(),
+                                             m_Context->getSourceManager()})
+        || !CO.IgnorePromptDiags /* XXX: raw input */)
+     return Result(D, true);
+
+    NamespaceDecl *NS = NamespaceDecl::Create(*m_Context, m_TU, /*inline=*/true, 
+                                              SourceLocation(), SourceLocation(),
+                                              &m_Context->Idents.get("__cling_N5"
+                                                + std::to_string((unsigned long)D)),
```

Comment:
This relies on us never re-using memory addresses. A counter is more stable and easier to debug (think breakpoint conditions)